src/ui/public/agg_types/metrics/scripted_metric.js

```diff
@@ -0,0 +1,38 @@
+import AggTypesMetricsMetricAggTypeProvider from 'ui/agg_types/metrics/metric_agg_type';
+import RegistryFieldFormatsProvider from 'ui/registry/field_formats';
+import textHtml from 'ui/agg_types/controls/text.html';
+
+export default function AggTypesMetricsScriptedMetricProvider(Private) {
+  let MetricAggType = Private(AggTypesMetricsMetricAggTypeProvider);
+  let fieldFormats = Private(RegistryFieldFormatsProvider);
+
+  return new MetricAggType({
+    name: 'scripted_metric',
+    title: 'Scripted Metric',
+    makeLabel: function () {
+      return 'Scripted Metric';
+    },
+    getFormat: function () {
+      return fieldFormats.getDefaultInstance('number') || fieldFormats.getDefaultInstance('percent');
+    },
+    params: [
+      {
+        name: 'init_script',
+        type: 'string',
+        editor: textHtml
+      }, {
+        name: 'map_script',
+        type: 'string',
+        editor: textHtml
+      }, {
+        name: 'combine_script',
+        type: 'string',
+        editor: textHtml
+      }, {
+        name: 'reduce_script',
+        type: 'string',
+        editor: textHtml
+      }
```

Comment:
I think that a global language selector is sufficient. Each param can specify a [`write()` method](https://github.com/elastic/kibana/blob/master/src/ui/public/agg_types/buckets/filters.js#L23-L44) that can produce the `{ "inline": "", "lang": "" }` version of the param:

```js
{
  name: 'reduce_script',
  type: 'string',
  write(aggConfig, output) {
    output.params.reduce_script = {
      lang: aggConfig.params.lang,
      inline: aggConfig.params.reduce_script
    }
  }
}
```